### PR TITLE
CDAP-5191 Adding modified FTPInputStream that doesn't throw an exception for trivial cases

### PR DESCRIPTION
**Pull request by gokulavasan**
This PR is required to add FTP Batch Source in hydrator-plugins.

JIRA : https://issues.cask.co/browse/CDAP-5191
Build : http://builds.cask.co/browse/CDAP-DUT3760

### CODE REVIEW

**Comment by gokulavasan** [comment]
@sreevatsanraman Please review.

File: cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.hadoop.fs.FSInputStream;
+import org.apache.hadoop.fs.FileSystem;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * FTP Input Stream, copied from Hadoop and modified, that doesn't thrown an exception when seeks are attempted to the
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java]: add link or fully qualified class name of the copied class. also typo: `doesn't throw an exception`

File: cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+ */
+@InterfaceAudience.Private
+@InterfaceStability.Unstable
+public class FTPInputStream extends FSInputStream {
+
+  InputStream wrappedStream;
+  FTPClient client;
+  FileSystem.Statistics stats;
+  boolean closed;
+  long pos;
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java]: maybe because you copied it, but these should be private?

Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java]: I didn't modify any code other than what was required.

**Comment by gokulavasan** [comment]
@gsps1 Addressed comment. I apologize, I squashed the commits before you could review.

File: cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    this.pos = 0;
+    this.closed = false;
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return pos;
+  }
+
+  // We don't support seek.
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java]: just update the comment here, as we now don't throw exception if ``getPos() == pos``

File: cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.hadoop.fs.FSInputStream;
+import org.apache.hadoop.fs.FileSystem;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * {@link FTPInputStream}, copied from Hadoop and modified, that doesn't throw an exception when seeks are attempted
+ * to the current position. Position equality check logic in {@link FTPInputStream#seek}  is the only change from the
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java]: nit : remove extra space before `is`

**Comment by gokulavasan** [comment]
@gsps1 Addressed comments. Created a JIRA as well (https://issues.cask.co/browse/CDAP-5387). Please review.

File: cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.hadoop.mapreduce.lib.input.LineRecordReader;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * {@link FTPInputStream}, copied from Hadoop and modified, that doesn't throw an exception when seeks are attempted
+ * to the current position. Position equality check logic in {@link FTPInputStream#seek} is the only change from the
+ * original class in Hadoop. This change is required since {@link LineRecordReader} calls {@link FTPInputStream#seek}
+ * with value of 0. This file can be removed once https://issues.cask.co/browse/CDAP-5387 is addressed.
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/org/apache/hadoop/fs/ftp/FTPInputStream.java]: add a todo tag.

**Comment by gsps1** [comment]
just one minor comment, squash and merge.

**Comment by gokulavasan** [comment]
Thanks @gsps1. Will merge after build passes.

**Comment by gokulavasan** [comment]
Build passed. Merging.

**Pull request merged**